keps/sig-node/5328-node-capabilities/README.md

```diff
@@ -0,0 +1,859 @@
+    // +optional 
+    // +featureGate=NodeCapabilities
+    Capabilities map[string]string `json:"capabilities,omitempty"`
+}
+
+
+// Node object remains unchanged in spec, only status is modified.
+type Node struct {
+    ...
+    Status NodeStatus `json:"status,omitempty" `
+}
+
+```
+
+**Example:**
+
+```
+Capabilities:
+   feature.kubernetes.io/inPlacePodResize: "true"
+   feature.kubernetes.io/podLevelResourcesSupport: "true"
+   feature.kubernetes.io/swapSupport: "true"
+   runtime.kubernetes.io/handlers: "runc, runsc"
+   kernel.kubernetes.io/LSMConfig: "seLinux"   
+```
+
+**Capability Reporting Semantics**
+
+1. If multiple interdependent settings are required to enable a feature, they must be collapsed into a single, logical capability. This simplifies the decision-making process for control plane components like the scheduler, which should not need to understand multiple interdependent settings. 
+For example, Swap support on a node depends on several conditions: the feature gate, Kubelet configuration, and underlying kernel support. Instead of reporting these three states individually, Kubelet evaluates them and advertises a single, simple capability: kubernetes.io/feature/swapSupport: "true".
+2. A capability should only be present if `node.status.capabilities` map if the feature is enabled and functional (`true` or another descriptive string if the capability is not boolean). If a feature is disabled (e.g., a beta feature gate is explicitly set to false) or not supported at all (e.g., on an older Kubelet version or an incompatible OS), the Kubelet must omit the corresponding capability key from the map entirely. This ensures there is consistency between a node that doesn't know about the feature and one that has it disabled which simplifies the logic for control plane components.
+3. The keys and values within the `node.status.capabilities` map must adhere to the same validation rules as standard Kubernetes labels. The Kubelet validates each capability and discards any key-value pair that is invalid, logging an error.
+
+**Note:**  
+
+*   We currently have [Node Features](https://github.com/kubernetes/api/blob/e8d4d542f6a9a16a694bfc8e3b8cd1557eecfc9d/core/v1/types.go#L6279) and [Node Runtime Features](https://github.com/kubernetes/api/blob/e8d4d542f6a9a16a694bfc8e3b8cd1557eecfc9d/core/v1/types.go#L6251) which publish some runtime features through Node Status. They are too narrowly scoped and currently not used for scheduling pods. We can deprecate those fields and introduce them into NodeCapabilities if required. 
+*   Other API options are explored in the [Alternatives](#alternatives) section.
+
+
+### kube-scheduler Changes
+
+To enable capability-based scheduling, this KEP proposes a new scheduler plugin named NodeCapabilityFilter. This plugin will be part of the default scheduling framework and will be enabled when the NodeCapabilities feature gate is active. DaemonSet Pods will be skipped by this plugin.
```

Comment:
Why deamonsets will be skipped? If we start using capabilities for scheduling a "correct" version of a Pod, I would think that two different DaemonSet may be needed for each LMSConfig?

Reply:
Ack. Removed.